bitsail-connectors/connector-doris/src/main/java/com/bytedance/bitsail/connector/doris/sink/proxy/DorisUpsertProxy.java

```diff
@@ -17,126 +17,101 @@
+
+import static com.bytedance.bitsail.connector.doris.sink.streamload.LoadStatus.PUBLISH_TIMEOUT;
+import static com.bytedance.bitsail.connector.doris.sink.streamload.LoadStatus.SUCCESS;
 
 public class DorisUpsertProxy extends AbstractDorisWriteModeProxy {
   private static final Logger LOG = LoggerFactory.getLogger(DorisUpsertProxy.class);
-
-  private static final long SCHEDULE_DELAY = 5000;
-
-  private List dorisBatchBuffers;
-  private static long lastFlushTimestamp;
-  private static long flushIntervalMs;
-  private long dorisBatchBuffersSize;
+  private static final List<String> DORIS_SUCCESS_STATUS = new ArrayList<>(Arrays.asList(SUCCESS, PUBLISH_TIMEOUT));
+  private final DorisExecutionOptions executionOptions;
+  private final DorisOptions dorisOptions;
+  private long lastCheckpointId = 0;
+  private final String labelPrefix;
+  private final LabelGenerator labelGenerator;
+  private final DorisWriterState dorisWriterState;
+  private final RecordStream recordStream;
 
   public DorisUpsertProxy(DorisExecutionOptions dorisExecutionOptions, DorisOptions dorisOptions) {
-    this.dorisExecutionOptions = dorisExecutionOptions;
-    this.dorisBatchBuffers = new ArrayList(dorisExecutionOptions.getBufferCount());
+    this.executionOptions = dorisExecutionOptions;
     this.dorisOptions = dorisOptions;
-    this.dorisStreamLoad = new DorisStreamLoad(dorisExecutionOptions, dorisOptions);
-    this.lastFlushTimestamp = System.currentTimeMillis();
-    this.flushIntervalMs = this.dorisExecutionOptions.getFlushIntervalMs();
-    this.dorisBatchBuffersSize = 0;
-    if (!this.dorisExecutionOptions.isBatch()) {
-      //flushIntervalMs only works in streaming mode
-      new Timer().schedule(new TimerTask() {
-        @Override
-        public void run() {
-          boolean flushIntervalReached = System.currentTimeMillis() - lastFlushTimestamp >= flushIntervalMs;
-          if (flushIntervalReached) {
-            LOG.info("Buffers is not fulfilled in {} ms", flushIntervalMs);
-            try {
-              addBatchBuffersOrFlush(null, true);
-            } catch (IOException e) {
-              e.printStackTrace();
-            }
-          }
-        }
-      }, SCHEDULE_DELAY, flushIntervalMs);
+    this.dorisWriterState = new DorisWriterState(executionOptions.getLabelPrefix());
+    this.labelPrefix = executionOptions.getLabelPrefix();
+    this.labelGenerator = new LabelGenerator(labelPrefix, dorisExecutionOptions.isEnable2PC());
+    this.recordStream = new RecordStream(executionOptions.getUpsertBufferSize(), executionOptions.getUpsertBufferCount());
+    init();
+  }
+
+  public void init() {
+    this.dorisStreamLoad = new DorisStreamLoad(executionOptions, dorisOptions, labelGenerator, recordStream);
+    try {
+      if (executionOptions.isEnable2PC()) {
+        dorisStreamLoad.abortPreCommit(labelPrefix, lastCheckpointId);
+      }
+      dorisStreamLoad.startLoad(labelGenerator.generateLabel(lastCheckpointId + 1));
```

Comment:
Why there is a `startLoad` during initialization?
I thought all `startLoad` are done during checkpoint.

Reply:
Since` prepareCommit` is executed earlier than `snapshotState`.
when initialization does not call `startLoad`, the `prepareCommit` calls `stopLoad` in first time , it will cause null pointer exception

Reply:
So `startLoad` is called but does not work during initialization? If so, I think it's ok.